File: go/launcher/proxy/screenshot.go
// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package screenshot includes a handler for the WebDriver screenshot operation.
package screenshot

import (
	"bytes"
	"context"
	"encoding/base64"
	"fmt"
	"image"
	"image/png"
	"net/http"

	"github.com/bazelbuild/rules_webtesting/go/launcher/proxy/driverhub"
	"github.com/bazelbuild/rules_webtesting/go/launcher/proxy/webdriver"
	"github.com/bazelbuild/rules_webtesting/go/util/cropper"
)

const sizeScript = `return {"width": screen.width, "height": screen.height};`

// ProviderFunc provides a handler for /screenshot command that crops the image if Chrome mobile emulation is enabled.
func ProviderFunc(session *driverhub.WebDriverSession, desired map[string]interface{}, base driverhub.HandlerFunc) (driverhub.HandlerFunc, bool) {
	chromeOptions, ok := desired["chromeOptions"].(map[string]interface{})
	if !ok {
		return base, false
	}

	if _, ok := chromeOptions["mobileEmulation"]; !ok {
		return base, false
	}

	return func(ctx context.Context, rq driverhub.Request) (driverhub.Response, error) {
		if rq.Method != http.MethodGet || len(rq.Path) != 1 || rq.Path[0] != "screenshot" {
			return base(ctx, rq)
		}
		img, err := session.Screenshot(ctx)
		if err != nil {
			body, _ := webdriver.MarshalError(err)
			return driverhub.Response{
				Status: webdriver.ErrorHTTPStatus(err),
				Body:   body,
			}, nil
		}

		val := struct {
			Width  int
			Height int
		}{}

		if err := session.ExecuteScript(ctx, sizeScript, nil, &val); err != nil {
			body, _ := webdriver.MarshalError(err)
			return driverhub.Response{
				Status: webdriver.ErrorHTTPStatus(err),
				Body:   body,
			}, nil
		}

		cropped, err := cropper.Crop(img, image.Rect(0, 0, val.Width, val.Height))
		if err != nil {
			body, _ := webdriver.MarshalError(err)
			return driverhub.Response{
				Status: webdriver.ErrorHTTPStatus(err),
				Body:   body,
			}, nil
		}

		buffer := &bytes.Buffer{}
		b64 := base64.NewEncoder(base64.StdEncoding, buffer)

		if err := png.Encode(b64, cropped); err != nil {
			body, _ := webdriver.MarshalError(err)
			return driverhub.Response{
				Status: webdriver.ErrorHTTPStatus(err),
				Body:   body,
			}, nil
		}

		b64.Close()

		return driverhub.Response{
			Status: http.StatusOK,
			Body:   []byte(fmt.Sprintf(`{"status": 0, "value": %q}`, buffer.String())),
		}, nil
	}, true
}

File: java/com/google/testing/screenshotter/Screenshot.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// //////////////////////////////////////////////////////////////////////////////
//
package com.google.testing.screenshotter;

import com.google.errorprone.annotations.Immutable;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Base64;
import javax.imageio.ImageIO;
import javax.imageio.ImageReader;
import org.json.JSONObject;

@Immutable
public final class Screenshot {
  private static final String FORMAT = "png";

  private final String base64;

  Screenshot(JSONObject response) {
    this.base64 = response.getString("value");
  }

  public String asBase64() {
    return base64;
  }

  public byte[] asBytes() {
    return Base64.getDecoder().decode(base64);
  }

  public BufferedImage asImage() throws IOException {
    ImageReader imageReader = ImageIO.getImageReadersByFormatName(FORMAT).next();
    imageReader.setInput(ImageIO.createImageInputStream(new ByteArrayInputStream(asBytes())), true);
    return imageReader.read(0);
  }

  @Override
  public boolean equals(Object other) {
    if (other == null) {
      return false;
    }
    if (!(other instanceof Screenshot)) {
      return false;
    }
    Screenshot o = (Screenshot) other;
    return asBase64().equals(o.asBase64());
  }

  @Override
  public int hashCode() {
    return asBase64().hashCode();
  }
}
